### ISSUE TITLE
Support for reproducible 'Documentation last generated' output

### ISSUE THREAD

**Issue by jayaddison** [sqlalchemyorg/zzzeeksphinx#23]
Hi folks - I was wondering whether a pull request to make the ['Documentation last generated'](https://github.com/sqlalchemyorg/zzzeeksphinx/blob/3d227af2a879cb5d66a17a31d1c477065755dd5a/zzzeeksphinx/themes/zsbase/layout.mako#L310) line reproducible (deterministic) on platforms that follow the relevant [Reproducible Builds standard](https://reproducible-builds.org/docs/source-date-epoch/) would be acceptable?

In supporting environments, that'd allow packages of `sqlalchemy`-plus-documentation to generate in a deterministic way, providing a bunch of [benefits and reassurances](https://reproducible-builds.org/docs/buy-in/) for downstream consumers.

(a patch to do so [has been suggested](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1042955) for downstream Debian; I'm offering to pick that up and provide it as a pull request here)

**Comment by zzzeek** [comment]
Hi -

I see what this is getting at, does sphinx itself support this?   Putting the time in UTC is one thing but it looks like this special (seemingly largely Debian centric-ish?) variable SOURCE_DATE_EPOCH seems to be a way to ..I guess provide a fixed value?   it's the "this is an arbitrary standard of unknown popularity/acceptance" that makes me hesistate here.  What does debian do for all the other packages that use sphinx docs?

**Comment by zzzeek** [comment]
the recommendation to put the git hash in the docs, rather than the timestamp, as indicated [here](https://reproducible-builds.org/docs/timestamps/) might be preferable.    That timestamp is there so that I personally can confirm the docs built on my site, not really any other reason.

**Comment by CaselIT** [comment]
I think assuming git is installed when generating the documentation would be fine.
The output could be something like
```
git show head --format='format:Generated on %cD from commit %H' -q
```
That prints
```
Generated on Fri, 16 Feb 2024 18:23:12 +0100 from commit 1c58fe53b6fd069cbb82955ddaf9eb5405076146
```

**Comment by CaselIT** [comment]
There is still the utc offset that may be different different.
Shame that there seems to be no way of telling git to use UTC timezone.

An option could be to tell it to use only the date, losing the time, or the unix timestamp making it hard to parse visually.
```
> git show head --format='format:Generated on %cs from commit %H' -q
Generated on 2024-02-16 from commit 1c58fe53b6fd069cbb82955ddaf9eb5405076146
```
```
> git show head --format='format:Generated on %ct from commit %H' -q
Generated on 1708104192 from commit 1c58fe53b6fd069cbb82955ddaf9eb5405076146

**Comment by zzzeek** [comment]
I think they want there to be no timestamp at all.  they are looking for the files diffing as identical across builds.   I think the point of SOURCE_DATE_EPOCH is that the build system sets this to a fixed number.

**Comment by CaselIT** [comment]
the timestamp is of the commit, so that would not change, unless the commit changes, and I don't see the use case for wanting a reproducible build with different commits

**Comment by zzzeek** [comment]
oh, timestamp of the commit is fine, sure

**Comment by CaselIT** [comment]
the only issue may be the timezone, unless git reports the timezone of the commit and not the local one. 

Interesting trying a bit it seems that it keeps the timezone of the committer, so the above line should be fine:

`git show head --format='format:Generated on %cD from commit %H' -q`

**Comment by jayaddison** [comment]
Good suggestions and details - it _is_ tricky to figure out a reliable build timestamp, especially across different systems, timezones, etc.  To respond in terms of abstraction layers: as SQLAlchemy is to database systems, `SOURCE_DATE_EPOCH` is to project build-time (it's an interface that can be implemented for any buildsystem out there; implementation details may vary by distro / CI system / makefile / ...).

**Comment by jayaddison** [comment]
Ah, and one of the benefits of the abstraction that I forgot to mention: because it's the responsibility of compatible buildsystems to make the `SOURCE_DATE_EPOCH` environment variable available during the build, there aren't any additional dependencies required in the code itself (in this case, in the `layout.mako` file in particular), regardless of whether the build is from a tarfile, `git` repository, SVN/RCS, or other origin.

Even so, the recommendation is to read `SOURCE_DATE_EPOCH` when available, but to fallback to the system clock (`time.time()` or similar) so that the project still builds in environments that don't configure the environment variable.

**Comment by CaselIT** [comment]
Seems like SOURCE_DATE_EPOCH is a standard variable https://reproducible-builds.org/docs/source-date-epoch/ .

Also it seems supported by sphinx, can't we just use that value from sphinx? https://github.com/sphinx-doc/sphinx/pull/1954

**Comment by jayaddison** [comment]
> Seems like SOURCE_DATE_EPOCH is a standard variable https://reproducible-builds.org/docs/source-date-epoch/ .

Yep - it is fairly well adopted and widely available across different systems (not specific to Debian, or any other tooling).

> Also it seems supported by sphinx, can't we just use that value from sphinx? https://github.com/sphinx-doc/sphinx/pull/1954

Re-using similar code makes sense I think - accessing it via `os.getenv`.  Sphinx doesn't re-export the value/timestamp itself via any public attributes/methods, but that's OK.

I could open a pull request to suggest the patch from downstream (and will include the original authorship info on the commit if so).

**Comment by zzzeek** [comment]
> Seems like SOURCE_DATE_EPOCH is a standard variable https://reproducible-builds.org/docs/source-date-epoch/ .
> 
> Also it seems supported by sphinx, can't we just use that value from sphinx? [sphinx-doc/sphinx#1954](https://github.com/sphinx-doc/sphinx/pull/1954)

that would be great but looking at that PR they only use it for "date", the current day, with no time.   so not very useful for my purposes (looking at the site, making sure docs got built).     

> not specific to Debian

it's not specific to Debian but overall the whole "reproducible builds" site / culture / initiative seems to be hosted by Debian.  which is fine.

the patch at https://bugs.debian.org/cgi-bin/bugreport.cgi?att=1;bug=1042955;filename=zzzeeksphinx.diff.txt;msg=5 is "fine", with two fixes:

1.utcfromtimestamp() is deprecated and warns:

```
DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).
```

2. the patch seems to switch from localtime (which is my timezone for me, since I build it on a host in my timezone :)  ) to UTC, which is fine, however, can as in 1 above use a timezone-aware datetime, then change the format string to add the timezone  `strftime("%c %Z")`.  thanks

**Comment by jayaddison** [comment]
How does this format seem?

![Created using Sphinx 7.2.6.  Documentation last generated: Thu 01 Jan 1970 00:00:00 UTC](https://github.com/sqlalchemyorg/zzzeeksphinx/assets/55152140/9c9e79d6-6dab-445d-9bd4-3cda97dc46cd)

**Comment by zzzeek** [comment]
yes that's great
